Memory Homework

bluemittens505 opened the pull request:
This code is mostly working, but it has a bug.  When you click on the same tile twice, the code considers the "tiles matching" and removes the tile.  I tried checking if this.$td and tile.$td were equal in Tile.prototype.matches(), but I didn't get the results I expected.  Is there a way to prevent this bug? Or is my whole design flawed? :-(

bluemittens505 committed:
Memory Homework

My first revision - still some kinks to work out

bluemittens505 committed:
Memory Homework

Has a bug.  When you click on the same tile twice, the code considers
the "tiles matching" and the tile disappears.

bluemittens505 committed:
Memory Homework

Fixed some errors displayed when I created the pull request.

afeld commented:
> is my whole design flawed?

Not at all – this looks really good!

> I tried checking if this.$td and tile.$td were equal in Tile.prototype.matches(), but I didn't get the results I expected

Ah. Remember in the first class when we were talking about different types in JavaScript, and how two objects that look the same aren't triple-equal (`===`)? For example:

``` javascript
var a = {};
var b = {};
a === b; // false
```

You ran into that exact issue here, because `clicked1.$td` is _not_ the same (jQuery) object as `clicked2.$td`, even though they contain the same element object under the hood (in a property). This happened because `$(this)` separately for each `Tile` that is created. Does that make sense? There are several ways to fix this, each with a simplified example:
1. Only create the jQuery object once.
   
   ``` javascript
   // change
   game.clicked1 = new Tile($(this));
   game.clicked2 = new Tile($(this));
   // to
   var $tile = $(this);
   game.clicked1 = new Tile($tile);
   game.clicked2 = new Tile($tile);
   ```
2. Compare the underlying element object rather than the jQuery object. Like an array, you can use square brackets to access the underlying elements from a jQuery object.
   
   ``` javascript
   // change
   clicked1.$td === clicked2.$td
   // to
   clicked1.$td[0] === clicked2.$td[0]
   ```
3. Use jQuery's [`.is()`](http://api.jquery.com/is/) method to compare two jQuery objects directly.
   
   ``` javascript
   // change
   clicked1.$td === clicked2.$td
   // to
   clicked1.$td.is(clicked2.$td)
   ```

Hope that helps? Remind me to bring this up in class – it's tricky!

bluemittens505 committed:
Memory Homework

Fixed bug with the help of Aidan.  Thanks.  Fixed cursor type on div
tags.  Re-factored some of the larger functions.  Still need to fix the
action().

bluemittens505 committed:
Memory Homework

Add missing semicolons

bluemittens505 committed:
Memory Homework

Add New Game button.  Refactored some more functions.

bluemittens505 committed:
Memory Homework

Moved most of code out of document.ready()

bluemittens505 committed:
Memory Homework

Re-arrange order of functions

bluemittens505 committed:
Revert "Memory Homework"

This reverts commit 38eec98a972ab2ebcf51301e11322537b79f0cce.

bluemittens505 committed:
Memory Homework

Final version on Tuesday night.  Hopefully more can be done tomorrow
night.

bluemittens505 committed:
Memory Homework

Added another class for buttons
Added another board to make sure all objects operate independently
More tweaks

bluemittens505 committed:
Memory Homework

More tweaks to pass Travis CI tests

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+var Game = function(boardElements) {
+	this.boardElements = boardElements;
+	this.tileClass = boardElements.tileHtmlElementClass.substr(1);
+	this.maxValue = boardElements.height*boardElements.width/2;
```
Don't be afraid of whitespace within lines :wink: Makes things more readable.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	var taken = true;
+	var index = this.takenValues.indexOf("#" + value + "#");
+	var lastIndex = this.takenValues.lastIndexOf("#" + value + "#");
+	if ( index === -1 || index === lastIndex ) {
```
This probably warrants an inline comment, to explain what this is checking for.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	this.remainingMatches = boardElements.height*boardElements.width/2;
+};
+
+Game.prototype.taken = function(value) {
```
Suggestion: methods that return `true`/`false` are often prefixed with `is`, i.e. `isTaken()`.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	this.boardElements = boardElements;
+	this.tileClass = boardElements.tileHtmlElementClass.substr(1);
+	this.maxValue = boardElements.height*boardElements.width/2;
+	this.takenValues = "#";
```
Why not use an array?

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	return taken;
+};
+
+Game.prototype.genValue = function() {
```
This name could be more descriptive...maybe `generateUniqueValue`?

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	$(self.boardElements.tileHtmlElementClass).click({
+		game:self
+	},action);
+};
```
I would keep this grouped with the other `Game.prototype` methods above.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	}
+	board += "</table>";
+	return board;
+};
```
You know, there might be enough logic in this to warrant its own class, e.g. a `BoardGenerator` or `TileValuesGenerator` or something. Then the `Game` would put all of it together and manage the actual tiles.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+
+Game.prototype.checkWinner = function() {
+	var self = this;
+	if ( --self.remainingMatches === 0 ) {
```
While this line works, it would be easier to read if it were split to two lines.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+var Tile = function($td) {
+	this.$td = $td;
+	this.$div = $td.find('div');
+	this.value = $td.find('div').text();
```
Minor: you could use `this.$div.text()` so that it doesn't need to search the DOM again.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	if ( this.value === tile.value ) {
+		match = true;
+	}
+	return match;
```
I would simplify this to

``` javascript
return this.value === tile.value;
```

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	return match;
+};
+
+Tile.prototype.showTile = function() {
```
Minor: it's implied that the method is acting on a `Tile`, so I would probably just name this `show`.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+
+Tile.prototype.sameTileAs = function($tile) {
+	return this.$td.is($tile);
+};
```
Since this is _essentially_ doing the same thing as `Tile.prototype.matches`, I would see if you can change your code to only need one or the other.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	return this.$td.is($tile);
+};
+
+var setClicked = function($tile) {
```
How could you avoid making this function global?

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+var action = function(event) {
+	var $tile = $(this);
+	var game = event.data.game;
+	if ( game.clicked1 === null ) {
```
Rather than checking for `null` specifically, it's more common to write `if ( !game.clicked1 )`.

afeld commented on memory.js:
```diff
@@ -0,0 +1,201 @@
+	});
+};
+
+var gameSetup = function(gameElements) {
```
Minor: I would change this to a verb, e.g. `setUpGame`.

afeld commented:
Left a bunch of nitpicky comments, but great work overall!
